service/runtime/tunnels.go

```diff
@@ -0,0 +1,155 @@
+package runtime
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/upvestco/httpsignature-proxy/service/logger"
+	"golang.org/x/exp/maps"
+)
+
+type UserCredentials struct {
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+}
+
+func (e UserCredentials) Empty() bool {
+	return len(e.ClientSecret) == 0 && len(e.ClientID) == 0
+}
+
+type Tunnels struct {
+	logger          logger.Logger
+	events          []string
+	closeGroup      *sync.WaitGroup
+	cancel          context.CancelFunc
+	tunnels         *tunnelsMap
+	logHeaders      bool
+	createApiClient func(credentials UserCredentials) ApiClient
+	proxyAddress    string
+}
+
+func CreateTunnels(logger logger.Logger, events []string, proxyAddress string, createApiClient func(credentials UserCredentials) ApiClient, logHeaders bool) *Tunnels {
+	return &Tunnels{
+		logger:          logger,
+		tunnels:         newTunnelsMap(),
+		events:          events,
+		closeGroup:      new(sync.WaitGroup),
+		createApiClient: createApiClient,
+		logHeaders:      logHeaders,
+		proxyAddress:    proxyAddress,
+	}
+}
+
+func (e *Tunnels) Stop() {
+	e.cancel()
+	if list := e.tunnels.list(); len(list) > 0 {
+		e.logger.Log("Closing webhooks tunnels")
+		for _, t := range list {
+			t.destroy()
+		}
+	}
+	e.closeGroup.Wait()
+}
+
+func AskForUserCredentials(proxyAddress string) (UserCredentials, int, error) {
+	client := http.Client{
+		Timeout: time.Second,
+	}
+	ctx := context.Background()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, proxyAddress+"/proxy-pass", nil)
+	if err != nil {
+		return UserCredentials{}, 0, errors.Wrap(err, "NewRequestWithContext")
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return UserCredentials{}, 0, errors.Wrap(err, "DefaultClient.Do")
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	if resp.StatusCode != http.StatusAccepted {
+		return UserCredentials{}, resp.StatusCode, nil
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return UserCredentials{}, 0, errors.Wrap(err, "ReadAll")
+	}
+	var uc UserCredentials
+
+	if err := json.Unmarshal(body, &uc); err != nil {
+		return UserCredentials{}, http.StatusBadRequest, nil
+	}
+	return uc, resp.StatusCode, nil
+}
+
+func (e *Tunnels) Start(userCredentialsCh chan UserCredentials) {
+	e.logger.Print(cyan("############################################################\n"))
+	e.logger.Print(cyan("To start listening webhook events - send /auth/token request\n"))
```

Comment:
lets change to: To start event listener, send an auth request: `POST /auth/token`

Reply:
Done